### DISCUSSION TITLE
How to resolve: FutureWarning: BLEDevice.metadata is deprecated and will be removed in a future version of Bleak, use AdvertisementData instead

### DISCUSSION THREAD

**Discussion by CalebDueck** [hbldh/bleak#1354] [Q&A] [answered]
* bleak version: 0.20.1
* Python version: 3.10.11
* Operating System: Windows 10
* BlueZ version (`bluetoothctl -v`) in case of Linux: NA

### Description

I'm using the metadata field but being forewarned that it is on its way to deprecation.
I am wondering how I could transition to AdvertisementData for this chunk of code

### What I Did

I read through https://github.com/hbldh/bleak/issues/1025 but I could see a clear answer as to how to use AdvertisementData as they mentioned "we could add an optional parameter" so I am unsure if that has been implemented already.

My code is currently:

```
devices = await BleakScanner.discover()
            for d in devices:
                if d.name == 'right_name':
                    right_name= d
                    sn = str(right_name.metadata['manufacturer_data'][65535])
                    sn = sn.replace("b'", "")
                    sn = sn.replace("'", "")
```
I'm wondering how I could iterate through devices and find their respective advertisement data for this.

### Logs

FutureWarning: BLEDevice.metadata is deprecated and will be removed in a future version of Bleak, use AdvertisementData instead

**Comment by dlech** [comment]
As described in the docs, the `discover()` method has a `return_adv` argument that changes the return type.

https://bleak.readthedocs.io/en/latest/api/scanner.html#bleak.BleakScanner.discover

**Reply by CalebDueck** [reply]
Thanks for directing me to the appropriate docs. 
I read through but am still having some issues.

I've updated my code to be the following:
```python
discovered_devices_and_advertisement_data = await BleakScanner.discover(return_adv=True)
            for key, dev_and_adv_dat in discovered_devices_and_advertisement_data.items():
                device = dev_and_adv_dat[0]
                adv_dat = dev_and_adv_dat[1]
                # print(d)
                if device.name == 'right_name':
                    right_name = device
                    sn = str(adv_dat['manufacturer_data'][65535])
```

I'm getting an tuple indices error on that last line.
I know that discovered_devices_and_advertisement_data is a tuple but from the docs it looks like AdvertisementData is a dictionary, not a tuple.

What would I need to change to make this work?

**Reply by dlech** [reply, marked as answer]
Can you give the full stack trace?


You can also simplify things a bit by changing

> ```python
>             for key, dev_and_adv_dat in discovered_devices_and_advertisement_data.items():
>                device = dev_and_adv_dat[0]
>                adv_dat = dev_and_adv_dat[1]
> ```

to 

```python
for device, adv_data in discovered_devices_and_advertisement_data.values():
```

and by changing `adv_dat['manufacturer_data']` to `adv_dat.manufacturer_data`

**Reply by CalebDueck** [reply]
Hey David, thanks for the two fixes.
the adv.manufacturer_data fixed my issue.

Much appreciated.

**Comment by dave-code-ruiz** [comment]
bleak version: 0.20.2
Python version: 3.11.4
Operating System: Raspbian OS with docker
BlueZ version (bluetoothctl -v) in case of Linux: 5.66

Hi, i use your code:

> 

        discovered_devices_and_advertisement_data = await BleakScanner.discover(return_adv=True)
        for device, adv_data in discovered_devices_and_advertisement_data.values():
            if device.address == address:
                self._device = device
                self._adv_data = adv_data

And I received an error:

> 
        for device, adv_data in discovered_devices_and_advertisement_data.values():
        AttributeError: 'list' object has no attribute 'values'

Please what am i doing wrong?

**Reply by dlech** [reply]
Perhaps you have a typo with the `return_adv` argument?

**Comment by dave-code-ruiz** [comment]
No, i havent 

Strange name for a typo ;)

**Comment by dave-code-ruiz** [comment]
I think funtion return List[BLEDevice] for overload function , but why?

**Reply by dlech** [reply]
If the argument `return_adv=True` is supplied, it should return a `dict`. Without the argument, it returns a `list` for backwards compatibility.

**Reply by dave-code-ruiz** [reply]
I don't know why , i supplied the argument but I receive List , and in log i see deprecated use of device.rssi.

**Comment by dave-code-ruiz** [comment]
Thanks for your time

**Comment by RobertWojtowicz** [comment]
@dlech 

Can you help with a similar case:
```
async def selectBLEdevices():
    print("Select your Omron device from the list below...")
    while(True):
        devices = await bleak.BleakScanner.discover()
        devices = sorted(devices, key = lambda x: x.rssi, reverse=True)
        tableEntries = []
        tableEntries.append(["ID", "MAC", "NAME", "RSSI"])
        for deviceIdx, device in enumerate(devices):
            tableEntries.append([deviceIdx, device.address, device.name, device.rssi])
        print(terminaltables.AsciiTable(tableEntries).table)
        res = input("Enter ID or just press Enter to rescan.\n")
        if(res.isdigit() and int(res) in range(len(devices))):
            break
    return devices[int(res)].address
```
Problem with:
```
devices = sorted(devices, key = lambda x: x.rssi, reverse=True)
```
and:
```
tableEntries.append([deviceIdx, device.address, device.name, device.rssi])
```

I get a return like this:
```
/root/omblepy/./omblepy.py:303: FutureWarning: BLEDevice.rssi is deprecated and will be removed in a future version of Bleak, use AdvertisementData.rssi instead
  devices = sorted(devices, key = lambda x: x.rssi, reverse=True)
/root/omblepy/./omblepy.py:307: FutureWarning: BLEDevice.rssi is deprecated and will be removed in a future version of Bleak, use AdvertisementData.rssi instead
  tableEntries.append([deviceIdx, device.address, device.name, device.rssi])
```

Full code: https://github.com/userx14/omblepy/blob/main/omblepy.py

BR,
Robert

**Reply by dlech** [reply]
Here you go:

```python
import asyncio
from collections import OrderedDict

from bleak import BleakScanner


async def main():
    devices = await BleakScanner.discover(return_adv=True)

    devices = OrderedDict(
        sorted(devices.items(), key=lambda x: x[1][1].rssi, reverse=True)
    )

    for i, (addr, (dev, adv)) in enumerate(devices.items()):
        print(i, addr, dev.name, adv.rssi)

asyncio.run(main())
```

Also, beware that `input()` can break your program because it is a long blocking operation. Bleak has background tasks running all of the time that need to be handled promptly when events are triggered. I would suggest using something like `aioconsole` instead.
